rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/ChargingRSpace.scala

```
@@ -119,5 +118,29 @@ object ChargingRSpace {
       override def createCheckpoint(): F[Checkpoint]    = space.createCheckpoint()
       override def reset(hash: Blake2b256Hash): F[Unit] = space.reset(hash)
       override def close(): F[Unit]                     = space.close()
+
+      implicit override val m: StorageMatch[F, BindPattern, ListParWithRandom] = space.m
+
+      override def getData(channel: Par): F[Seq[internal.Datum[ListParWithRandom]]] =
```

ah, yes - those methods were properly hidden from rholang by the PureRSpace. I guess thought the indirection and the confusion it caused are not worth it, assuming we all agree those should not be used.

---

Yeah. This could be ameliorated if `ISpace` was broken into a number of smaller traits.

---

Instead of agreeing that this should not be used why not make sure this is the case by failing fast with `???` instead of delegating?

---

I've just checked, that the only methods rholang needs are: `install`, `consume`, `produce`, `close`. @lukasz-golebiewski @dzajkowski would you be OK with introducing a `Tuplespace` trait containing them, that the ISpace trait would extend?

---

Sure thing

---

@JosDenmark do you want to take care of it?

---

Happy to.